designs/2025-rule-tester-assertion-options/README.md

```diff
@@ -0,0 +1,358 @@
+The options could be defined on two levels. On `RuleTester`'s `constructor` effectively impacting all tests, or on the test `run` method itself, so only that set is affected.
+
+## Detailed Design
+
+### Variant 1 - Constructor based options
+
+```ts
+new RuleTester(testerConfig: {...}, assertionOptions: {
+  /**
+   * Require message assertion for each invalid test case.
+   * If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+   * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+   * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+   *
+   * @default false
+   */
+  requireMessage: boolean | 'message' | 'messageId';
+  /**
+   * Require full location assertions for each invalid test case.
+   * If `true`, errors must extend `Array<{ line: number, column: number, endLine?: number | undefined, endColumn?: number | undefined }>`.
+   * `endLine` or `endColumn` may be absent, if the observed error does not contain these properties.
+   *
+   * @default false
+   */
+  requireLocation: boolean;
+} = {});
+```
+
+### Variant 2 - Test method based options
+
+```ts
+ruleTester.run("rule-name", rule, {
+  assertionOptions?: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+     * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+     * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+     *
+     * @default false
+     */
+    requireMessage?: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, errors must extend `Array<{ line: number, column: number, endLine?: number | undefined, endColumn?: number | undefined }>`.
+     * `endLine` or `endColumn` may be absent or undefined, if the observed error does not contain these properties.
+     *
+     * @default false
+     */
+    requireLocation?: boolean;
+  },
+  valid: [ /* ... */ ],
+  invalid: [ /* ... */ ],
+});
+```
+
+### Shared Logic
+
+#### requireMessage
+
+If `requireMessage` is set to `true`, the invalid test case cannot consist of an error count assertion only, but must also include a message assertion. (See below)
+This can be done either by providing only a `string`/`RegExp` message, or by using the `message`/`messageId` property of the error object in the `TestCaseError` (Same as the current behavior).
+If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
```

Comment:
As above, this could use some clarification.

Reply:
Updated